### PR TITLE
Stub out DCM connection for core, memory requests

### PR DESCRIPTION
**Pull request by hunhoffe**
This pull request includes what is necessary to run requests for memory, core from a client to the controller to the DCM scheduler instance. While resources are not yet actually allocated, the general framework and communication channels are functional.

As part of this, the changes include:
- Separating rackscale syscall code into different sets of operations (e.g., a folder for process operations, a folder for file operations)
- Modifying run.py so you can setup the network separately from build/run. This was necessary because DCM needs the network setup, but DCM also needs to be started before the controller starts running... so there were previously timing issues between starting the controller and DCM
- Modify run.py to fetch/use releases of the nrk-dcm-scheduler repo. 
- Update all integration tests for the rackscale/exokernel to use DCM correctly, and create new DCM integration test
- Removes client_shmem and server_shmem. To do so, created more comprehensive RPC transport API which makes it easier to choose the most efficient way to send informatio (e.g., as single buffer when possible for reduced number of reads/write).

### CODE REVIEW

File: lib/vmxnet3/src/vmx.rs [hunk 1]
```diff
@@ -48,6 +48,8 @@ struct TRXQueueShared {
     buffer: *mut u8,
 }
 
+unsafe impl Send for TRXQueueShared {}
```
Comment (hunhoffe) [inline lib/vmxnet3/src/vmx.rs:51]: @gz I forgot we never touched base again for whether this is truly safe or not

File: lib/rpc/src/client.rs [hunk 2]
```diff
@@ -48,72 +48,18 @@ impl RPCClient for Client {
...
     ) -> Result<(), RPCError> {
         // Calculate total data_out len
-        let data_out_len = data_out.iter().fold(0, |acc, x| acc + x.len());
         let data_in_len = data_in.iter().fold(0, |acc, x| acc + x.len());
 
-        // Check lengths
-        assert!(data_out_len + HDR_LEN <= self.transport.max_send());
-        assert!(data_in_len + HDR_LEN <= self.transport.max_recv());
+        // Create request header and send message
+        let mut hdr = unsafe { &mut *self.hdr.get() };
```
Comment (gz) [inline lib/rpc/src/client.rs:54]: Maybe add Safety comment why it's safe to create a mutable reference here

File: lib/rpc/src/server.rs [hunk 3]
```diff
@@ -15,79 +14,47 @@ use crate::transport::Transport;
...
-
-        {
-            let hdr = self.hdr.borrow();
-            let total_msg_data = hdr.msg_len as usize;
-            let mut buff = self.buff.borrow_mut();
-            self.transport.recv(&mut buff[..total_msg_data]).unwrap();
-        }
-        Ok(self.hdr.borrow().msg_type)
+        self.transport.recv_mbuf(unsafe { &mut *self.mbuf.get() })?;
+        Ok(unsafe { (*self.mbuf.get()).hdr.msg_type })
```
Comment (gz) [inline lib/rpc/src/server.rs:40]: Same here, add safety comment

File: lib/rpc/src/server.rs [hunk 4]
```diff
@@ -15,79 +14,47 @@ use crate::transport::Transport;
...
+            .try_recv_mbuf(unsafe { &mut *self.mbuf.get() })?
         {
-            let hdr = self.hdr.borrow();
-            let total_msg_data = hdr.msg_len as usize;
-            let mut buff = self.buff.borrow_mut();
-            self.transport.recv(&mut buff[..total_msg_data]).unwrap();
+            return Ok(None);
         }
-        Ok(Some(self.hdr.borrow().msg_type))
+        Ok(Some(unsafe { (*self.mbuf.get()).hdr.msg_type }))
```
Comment (gz) [inline lib/rpc/src/server.rs:52]: see above

File: lib/rpc/src/server.rs [hunk 5]
```diff
@@ -15,79 +14,47 @@ use crate::transport::Transport;
...
     fn reply(&self) -> Result<(), RPCError> {
-        // Send response header + data
-        let hdr = self.hdr.borrow();
-        let msg_len = hdr.msg_len as usize;
-        let hdr_slice = unsafe { hdr.as_bytes() };
-        self.transport.send(hdr_slice)?;
-
-        let buff = self.buff.borrow_mut();
-        self.transport.send(&buff[0..msg_len])
+        self.transport.send_mbuf(unsafe { &mut *self.mbuf.get() })
```
Comment (gz) [inline lib/rpc/src/server.rs:57]: same

File: lib/rpc/src/server.rs [hunk 6]
```diff
@@ -116,7 +83,9 @@ impl<'a> RPCServer<'a> for Server<'a> {
         self.receive()?;
 
         // TODO: registration
-        let client_id = func(&mut self.hdr.borrow_mut(), &mut self.buff.borrow_mut())?;
+        let client_id = func(unsafe { &mut (*self.mbuf.get()).hdr }, unsafe {
```
Comment (gz) [inline lib/rpc/src/server.rs:86]: same

File: lib/rpc/src/server.rs [hunk 7]
```diff
@@ -131,8 +100,9 @@ impl<'a> RPCServer<'a> for Server<'a> {
         match self.handlers.borrow().get(&rpc_id) {
             Some(func) => {
                 {
-                    let mut hdr = self.hdr.borrow_mut();
-                    func(&mut hdr, &mut self.buff.borrow_mut())?;
+                    func(unsafe { &mut (*self.mbuf.get()).hdr }, unsafe {
```
Comment (gz) [inline lib/rpc/src/server.rs:103]: same

File: lib/rpc/src/server.rs [hunk 8]
```diff
@@ -149,8 +119,9 @@ impl<'a> RPCServer<'a> for Server<'a> {
             Some(rpc_id) => match self.handlers.borrow().get(&rpc_id) {
                 Some(func) => {
                     {
-                        let mut hdr = self.hdr.borrow_mut();
-                        func(&mut hdr, &mut self.buff.borrow_mut())?;
+                        func(unsafe { &mut (*self.mbuf.get()).hdr }, unsafe {
```
Comment (gz) [inline lib/rpc/src/server.rs:122]: same

**Review by gz — APPROVED** [review]

**Pull request merged**